implementation of `tuple_element<i,camp::tuple<T...>>` not allowing structured binding using `const auto`

Hello,
In the StructuredBindings unit test, if I change the return type to `const auto` as such:
```c++
TEST(CampTuple, StructuredBindings)
{
  auto t = camp::make_tuple(3, 9.9);
  const auto [a, b] = t;
  ASSERT_EQ(a, 3);
  ASSERT_NEAR(b, 9.9, 1e-15);
}
```
I get the following error:
```
[ 34%] Building CXX object test/CMakeFiles/tuple.dir/tuple.cpp.o
cd /usr/workspace/settgast/Codes/geosx/camp/build/test && /usr/lib64/ccache/c++ -DGTEST_HAS_DEATH_TEST=0 -I/usr/workspace/settgast/Codes/geosx/camp/include -I/usr/workspace/settgast/Codes/geosx/camp/build/include -isystem /usr/workspace/settgast/Codes/geosx/camp/extern/blt/thirdparty_builtin/googletest/googletest/include -Wall -Wextra      -fPIE -std=c++17 -MD -MT test/CMakeFiles/tuple.dir/tuple.cpp.o -MF CMakeFiles/tuple.dir/tuple.cpp.o.d -o CMakeFiles/tuple.dir/tuple.cpp.o -c /usr/workspace/settgast/Codes/geosx/camp/test/tuple.cpp
/usr/workspace/settgast/Codes/geosx/camp/test/tuple.cpp: In member function ‘virtual void CampTuple_StructuredBindings_Test::TestBody()’:
/usr/workspace/settgast/Codes/geosx/camp/test/tuple.cpp:362:23: error: binding reference of type ‘int&&’ to ‘camp::tuple_element_t<0, camp::tuple<int, double> >’ {aka ‘const int’} discards qualifiers
   const auto [a, b] = t;
                       ^
/usr/workspace/settgast/Codes/geosx/camp/test/tuple.cpp:362:23: error: binding reference of type ‘double&&’ to ‘camp::tuple_element_t<1, camp::tuple<int, double> >’ {aka ‘const double’} discards qualifiers
```

If I change the `camp::tuple` to `std::tuple`, it compiles fine.

Also...in tuple.hpp if I replace the implementation:
```c++
  template <size_t i, typename ... T>
  struct tuple_element<i, camp::tuple<T...>> {
    using type = decltype(camp::get<i>(camp::tuple<T...>{}));
  };
```
with something I took/altered from [cppreference](https://en.cppreference.com/w/cpp/utility/tuple/tuple_element):
```c++
  // recursive case
  template<std::size_t I, class Head, class... Tail>
  struct tuple_element<I, camp::tuple<Head, Tail...>>
      : std::tuple_element<I - 1, camp::tuple<Tail...>>
  { };
 
// base case
  template<class Head, class... Tail>
  struct tuple_element<0, camp::tuple<Head, Tail...>>
  {
      using type = Head;
  };
```
it compiles fine.

@trws @adayton1 Is this an OK change to make? I can submit a PR with the changes and add a test for the const case if so. Also should probably add testing for all the structured binding cases: `const auto`, `const auto &`, `const auto &&`?

--- a/include/camp/tuple.hpp
+++ b/include/camp/tuple.hpp
@@ -703,7 +703,7 @@ namespace std {
 
   template <size_t i, typename ... T>
   struct tuple_element<i, camp::tuple<T...>> {
-    using type = decltype(camp::get<i>(camp::tuple<T...>{}));
+    using type = ::camp::tuple_element_t<i, camp::tuple<T...>>;
   };
 
   /// This allows structured bindings to be used with camp::tagged_tuple
@@ -718,7 +718,7 @@ namespace std {
 
   template <size_t i, typename TagList, typename... Elements>
   struct tuple_element<i, camp::tagged_tuple<TagList, Elements...>> {
-    using type = decltype(camp::get<i>(camp::tagged_tuple<TagList, Elements...>{}));
+    using type = ::camp::tuple_element_t<i, camp::tagged_tuple<TagList, Elements...>>;
   };
 } // namespace std
 #endif

--- a/test/tuple.cpp
+++ b/test/tuple.cpp
@@ -360,8 +360,8 @@ TEST(CampTuple, StructuredBindings)
 {
   auto t = camp::make_tuple(3, 9.9);
   auto [a, b] = t;
-  static_assert(std::is_same_v< decltype(a), int > );
-  static_assert(std::is_same_v< decltype(b), double > );
+  static_assert(std::is_same< decltype(a), int >::value );
+  static_assert(std::is_same< decltype(b), double >::value );
   ASSERT_EQ(a, 3);
   ASSERT_NEAR(b, 9.9, 1e-15);
 }

--- a/test/tuple.cpp
+++ b/test/tuple.cpp
@@ -360,9 +360,53 @@ TEST(CampTuple, StructuredBindings)
 {
   auto t = camp::make_tuple(3, 9.9);
   auto [a, b] = t;
+  static_assert(std::is_same_v< decltype(a), int > );
+  static_assert(std::is_same_v< decltype(b), double > );
   ASSERT_EQ(a, 3);
   ASSERT_NEAR(b, 9.9, 1e-15);
 }
+
+TEST(CampTuple, StructuredBindingsConst)
+{
+  auto t = camp::make_tuple(3, 9.9);
+  const auto [a, b] = t;
+  static_assert(std::is_same_v< decltype(a), const int > );
+  static_assert(std::is_same_v< decltype(b), const double > );
+  ASSERT_EQ(a, 3);
+  ASSERT_NEAR(b, 9.9, 1e-15);
+}
+
+TEST(CampTuple, StructuredBindingsRef)
+{
+  auto t = camp::make_tuple(3, 9.9);
+  auto & [a, b] = t;
+  static_assert(std::is_same< decltype(a), int >::value ); // Not an int &
+  static_assert(std::is_same< decltype(b), double >::value ); // Not a double &
+  std::cout<<typeid(a).name()<<std::endl;
+  ASSERT_EQ(a, 3);
+  ASSERT_NEAR(b, 9.9, 1e-15);
+
+  a = 4; // Well...kind of an int &
+  b = 10.1; // also kind of a double &
+  ASSERT_EQ(camp::get<0>(t), 4);
+  ASSERT_NEAR(camp::get<1>(t), 10.1, 1e-15);
+}
+
+TEST(CampTuple, StructuredBindingsRefConst)
+{
+  auto t = camp::make_tuple(3, 9.9);
+  const auto & [a, b] = t;
+  static_assert(std::is_same< decltype(a), const int >::value );  // Not an const int &
+  static_assert(std::is_same< decltype(b), const double >::value ); // Not a const double &
+  std::cout<<typeid(a).name()<<std::endl;
+  ASSERT_EQ(a, 3);
+  ASSERT_NEAR(b, 9.9, 1e-15);
+
+  // these should fail to compile.
+  // a = 4;
+  // b = 10.1;
+}
+
 #endif
 
 struct s1;

--- a/test/tuple.cpp
+++ b/test/tuple.cpp
@@ -382,7 +382,6 @@ TEST(CampTuple, StructuredBindingsRef)
   auto & [a, b] = t;
   static_assert(std::is_same< decltype(a), int >::value ); // Not an int &
   static_assert(std::is_same< decltype(b), double >::value ); // Not a double &
-  std::cout<<typeid(a).name()<<std::endl;
   ASSERT_EQ(a, 3);
   ASSERT_NEAR(b, 9.9, 1e-15);
 
@@ -398,15 +397,9 @@ TEST(CampTuple, StructuredBindingsRefConst)
   const auto & [a, b] = t;
   static_assert(std::is_same< decltype(a), const int >::value );  // Not an const int &
   static_assert(std::is_same< decltype(b), const double >::value ); // Not a const double &
-  std::cout<<typeid(a).name()<<std::endl;
   ASSERT_EQ(a, 3);
   ASSERT_NEAR(b, 9.9, 1e-15);
-
-  // these should fail to compile.
-  // a = 4;
-  // b = 10.1;
 }
-
 #endif
 
 struct s1;

--- a/test/tuple.cpp
+++ b/test/tuple.cpp
@@ -370,8 +370,8 @@ TEST(CampTuple, StructuredBindingsConst)
 {
   auto t = camp::make_tuple(3, 9.9);
   const auto [a, b] = t;
-  static_assert(std::is_same_v< decltype(a), const int > );
-  static_assert(std::is_same_v< decltype(b), const double > );
+  static_assert(std::is_same< decltype(a), const int >::value );
+  static_assert(std::is_same< decltype(b), const double >::value );
   ASSERT_EQ(a, 3);
   ASSERT_NEAR(b, 9.9, 1e-15);
 }

